### PR TITLE
Allows additional payment checks in payment method list

### PR DESCRIPTION
**Pull request by jensscherbl**
As discussed with @okorshenko at the Magento Contribution Days 2019 in Munich, this backwards compatible change allows developers to whitelist additional payment checks in [Payment/Model/MethodList.php][1].

```xml
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <type name="Magento\Payment\Model\Checks\SpecificationFactory">
        <arguments>
            <argument name="mapping" xsi:type="array">
                <item name="acme_custom_payment_method_check" xsi:type="object">
                    Acme\Payment\Model\Checks\CustomPaymentMethodCheck
                </item>
            </argument>
        </arguments>
    </type>
    <type name="Magento\Payment\Model\MethodList">
        <arguments>
            <argument name="additionalChecks" xsi:type="array">
                <item name="acme_custom_payment_method_check" xsi:type="string">
                    acme_custom_payment_method_check
                </item>
            </argument>
        </arguments>
    </type>
</config>
```

This is already implemented in the same way in other classes like [Payment\Block\Form\Container][2], by the way.

For a proper use case and additional explanation, please check out [smaex/customer-group-payments][3], which at the moment requires [smaex/additional-payment-checks][4] as a workaround.

[1]: https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Payment/Model/MethodList.php#L80
[2]: https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Payment/Block/Form/Container.php#L93
[3]: https://github.com/smaex/customer-group-payments
[4]: https://github.com/smaex/additional-payment-checks

### CODE REVIEW

File: app/code/Magento/Payment/Model/MethodList.php [hunk 1]
```diff
@@ -39,19 +39,29 @@ class MethodList
...
      * @param \Magento\Payment\Helper\Data $paymentHelper
-     * @param Checks\SpecificationFactory $specificationFactory
+     * @param Checks\SpecificationFactory  $specificationFactory
+     * @param array                        $additionalChecks
      */
     public function __construct(
         \Magento\Payment\Helper\Data $paymentHelper,
-        \Magento\Payment\Model\Checks\SpecificationFactory $specificationFactory
+        \Magento\Payment\Model\Checks\SpecificationFactory $specificationFactory,
+        array $additionalChecks = []
```
Comment (orlangur) [inline app/code/Magento/Payment/Model/MethodList.php:55]: It seems a bit awkward to me to inject two dependencies: `$specificationFactory` and its additional configuration in `$additionalChecks`.

Why not pass some `$methodApplicabilityChecks` with default value containing four predefined checks?

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/Payment/Model/MethodList.php (hunk 1 above)
Comment (jensscherbl) [reply app/code/Magento/Payment/Model/MethodList.php:55]: That was exactly my first thought as well. The factory creates a composite check containing the defined checks, so a simpler implementation would be to just inject different composite checks in different places, and add custom checks to these composite checks where needed.

However, I already discussed this idea with @okorshenko, and he wanted to keep the current whitelisting approach for backwards compatibility reasons. Again, this is how Magento already does it in other places, like [Payment/Block/Form/Container][1].

[1]: https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Payment/Block/Form/Container.php#L93

**Review by larsroettig — APPROVED** [review]
I think we need to move the Refactoring in an extra Ticket this Pull request help to change the current bad situation.

**Review by VladimirZaets — CHANGES_REQUESTED** [review]
Hi @jensscherbl. Thanks for collaboration. Due to [Magento Definition of Done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) all code must be covered by tests. Please cover your fix by automated tests.

**Comment by VladimirZaets** [comment]
@jensscherbl, I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for collaboration

**Comment by jensscherbl** [comment]
@VladimirZaets  

> I am closing this PR now due to inactivity.

That‘s ok. We just leave it broken instead of merging the working fix, so other developers get the chance to run into this frustrating experience as well. 👍

**Comment by jensscherbl** [comment]
> I am closing this PR now due to inactivity.

@VladimirZaets That‘s ok. Just curious, what are you gonna do about the issue instead?

**Comment by VladimirZaets** [comment]
Hi @jensscherbl. We covered this PR by tests, but we can't process it until you don't sign CLA. Can you please sign CLA and we will process this PR.

**Comment by jensscherbl** [comment]
@VladimirZaets 

> Can you please sign CLA and we will process this PR.

Done. Thanks.

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-6816](https://jira.corp.magento.com/browse/ENGCOM-6816) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
@jensscherbl thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Pull request merged**